n2survey/lime/survey.py

```diff
@@ -158,6 +159,76 @@ def read_responses(self, responses_file: str) -> None:
+            ValueError: Unknown question types.
+
+        Returns:
+            [pd.DataFrame]: The response for the selected question.
+        """
+        question_group = self.questions[self.questions.question_group == question]
+        question_types = np.unique(question_group.type)
+        question_responses = self.responses.loc[:, question_group.index]
+        if len(question_types) > 1:
+            raise ValueError(f'Question {question} has multiple types {question_types}.')
+        else:
+            question_type = question_types[0]
+
+        if question_type == 'single-choice':
+            # ASSUME: question_responses consists only one column
+            #         in result question_group also consists of only one item
+            # create a DataFrame because questions_response is only a Series
+            response = pd.DataFrame()
+            response[question] = question_responses[question_responses.columns[0]]
+            if labels:
+                # rename column and the column entries
+                response = response.rename(columns=dict(question_group.label))
+                for col_idx in range(len(response.columns)):
+                    response.iloc[:, col_idx] = response.iloc[:, col_idx] \
+                        .cat.rename_categories(question_group.choices[col_idx])
+        elif question_type == 'multiple-choice':
+            # ASSUME: question response consists of multiple columns with
+            #         'Y' or NaN as entries.
+            response = question_responses.notnull()
+            if labels:
+                # WARNING: If choice has no possible mapping, it will be skipped.
+                # rename columns only because entries are True or False
+                response = response.rename(
+                    columns=dict(
+                        [(x, y['Y']) for x, y in question_group.choices.items()
+                         if y is not np.nan]
+                    )
+                )
+        elif question_type == 'free':
+            # ASSUME: question_responses consists of only one column with
+            #         unlimited options of answers.
+            response = question_responses
+            if labels:
+                # rename only columns because entries are not categorical
+                response = response.rename(columns=dict(question_group.label))
+        elif question_type == 'array':
+            # ASSUME: question_responses consists of multiple columns with
+            #         each column has the same set of possible choices
+            response = question_responses
+            if labels:
+                # rename all columns and their entries
+                response = response.rename(columns=dict(question_group.label))
+                for col_idx in range(len(response.columns)):
+                    response.iloc[:, col_idx] = response.iloc[:, col_idx] \
+                        .cat.rename_categories(question_group.choices[col_idx])
+        else:
```

Comment:
Most of this code is not needed, if `labels == False`. So, I would suggest restructuring the function it a bit:


```python

# Get Sub-DataFrame of the corresponding responses
question_responses = ...

# Change values to bool, if it is multiple choice
if question_type == 'multiple-choice':
    question_responses = question_responses.notnull()

# Apply labels
if labels:
    # Rename column names
    if question_type == 'multiple-choice':
        # Rename column names
        # New method `self.get_choices(question)` returns dict like `{<column name>: <"Y" choice label>}`
        # NOTE: You might need to add something for a contingent question
         question_responses= question_responses.rename(columns=self.get_choices(question))
    else:
        # Rename values
        # Adding `question_group[column].choices is not None` to the loop handles all cases:
        #   it will work for array, for a single choice question, and also will skip contingent question
        for column in  question_responses.columns:
            if question_group[column].choices is not None:
                question_responses.loc[:, column] = question_responses.loc[:, column] \
                        .cat.rename_categories(question_group[column].choices)

        # Rename column names
        question_responses= question_responses.rename(columns=dict(question_group.label))
          

return question_responses 
```